### DISCUSSION TITLE
[Request Help] How to create CSR with key on smart card

### DISCUSSION THREAD

**Discussion by PIKACHUIM** [OpenSC/OpenSC#3233] [Q&A] [answered]
Hello, thank you for providing this tool
I have a question that I hope to receive an answer to:
How to create a certificate request using the key generated on the smart card and sign it using the key on the smart card in the application of X509 certificate using OpenSC
I have seen some examples using OpenSSL, but encountered issues:

```
openssl.exe engine dynamic -pre SO_PATH:".\libp11\32bit\pkcs11.dll" -pre ID:pkcs11 -pre LIST_ADD:1 -pre LOAD -pre MODULE_PATH:".\pkcs11\opensc-pkcs11.dll"
(dynamic) Dynamic engine loading support
[Success]: SO_PATH:.\libp11\32bit\pkcs11.dll
[Success]: ID:pkcs11
[Success]: LIST_ADD:1
[Failure]: LOAD
BCAD0000:error:12800067:DSO support routines:win32_load:could not load the shared library:..\crypto\dso\dso_win32.c:108:filename(.\libp11\32bit\pkcs11.dll)
BCAD0000:error:12800067:DSO support routines:DSO_load:could not load the shared library:..\crypto\dso\dso_lib.c:152:
BCAD0000:error:13000084:engine routines:dynamic_load:dso not found:..\crypto\engine\eng_dyn.c:442:
[Failure]: MODULE_PATH:.\pkcs11\opensc-pkcs11.dll
BCAD0000:error:13000089:engine routines:int_ctrl_helper:invalid cmd name:..\crypto\engine\eng_ctrl.c:90:
BCAD0000:error:13000089:engine routines:ENGINE_ctrl_cmd_string:invalid cmd name:..\crypto\engine\eng_ctrl.c:258:
```

System: Windows, TPM virtual smart card
I can now create X509 certificates externally (such as OpenSSL) and import them, 
but I don't know how to create a key on the smart card and use it to create the certificate

I hope to receive an answer. Thank you again

**Comment by Jakuje** [comment]
Hi. First of all, I moved this to discussion, which is the place where questions should come, unlike the issues where the bugs, rfes and coding/release related stuff should come.

To your issue, there are different tools you can make CSR with private key on the card/pkcs11. I have from top of my head example that I used in p11test to generate certificates using gnutls p11tool:

https://github.com/OpenSC/OpenSC/blob/master/src/tests/p11test/runtest.sh#L73

Or similarly here:

https://github.com/latchset/pkcs11-provider/blob/main/tests/setup.sh#L98

Your error says that the pkcs11 engine is not installed in the expected location. Given that the engines are deprecated in current OpenSSL, I do not think it makes much sense to try to resurrect this use case. But if you want to use the openssl, I would suggest to use the pkcs11-provider, but I do not have any ready-made command for now.

**Comment by PIKACHUIM** [comment]
Thank you for your prompt assistance. I will try the tutorial you provided.

btw, I want to know why OpenSC cannot officially provide CSR generation function. 
Since OpenSC already provides the function of creating keys and signatures using smart cards.

Thank you again for your kind help!

**Reply by Jakuje** [reply]
The OpenSC is the driver for the smart cards and pkcs11-tool is a low-level tool to handle the smart cards operations. Implementing all the corner cases of the X.509 certificates that involve  CSR creation, signing and others would be a significant amount of work nobody did yet, because, for example, there are different tools that can do that better (certtool, openssl, ...). If you believe it should be part of pkcs11-tool or any different tool part of the opensc, we welcome every contribution.

**Reply by PIKACHUIM** [reply, marked as answer]
Yes, I found that on Windows, you can directly use Certreq to create certificate requests using the key on the smart card. 
At the same time, it can use OpenSC's minidriver as the driver module, just like this:
```
certreq -new -f request.inf request.csr
```
The `request.inf` is sourced from [certreq](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/certreq_1)
Thank you for your response!
